src/newgrf_roadstop.cpp

```diff
@@ -0,0 +1,614 @@
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file command.cpp Handling of NewGRF road stops. */
+
+#include "stdafx.h"
+#include "debug.h"
+#include "station_base.h"
+#include "roadstop_base.h"
+#include "newgrf_roadstop.h"
+#include "newgrf_class_func.h"
+#include "newgrf_cargo.h"
+#include "newgrf_roadtype.h"
+#include "gfx_type.h"
+#include "company_func.h"
+#include "road.h"
+#include "window_type.h"
+#include "date_func.h"
+#include "town.h"
+#include "viewport_func.h"
+#include "newgrf_animation_base.h"
+#include "newgrf_sound.h"
+
+#include "safeguards.h"
+
+template <typename Tspec, typename Tid, Tid Tmax>
+void NewGRFClass<Tspec, Tid, Tmax>::InsertDefaults()
+{
+	/* Set up initial data */
+	classes[0].global_id = 'DFLT';
+	classes[0].name = STR_STATION_CLASS_DFLT;
+	classes[0].Insert(nullptr);
+
+	classes[1].global_id = 'WAYP';
+	classes[1].name = STR_STATION_CLASS_WAYP;
+	classes[1].Insert(nullptr);
+}
+
+template <typename Tspec, typename Tid, Tid Tmax>
+bool NewGRFClass<Tspec, Tid, Tmax>::IsUIAvailable(uint index) const
+{
+	return true;
+}
+
+INSTANTIATE_NEWGRF_CLASS_METHODS(RoadStopClass, RoadStopSpec, RoadStopClassID, ROADSTOP_CLASS_MAX)
+
+static const uint NUM_ROADSTOPSPECS_PER_STATION = 63; ///< Maximum number of parts per station.
+
+uint32 RoadStopScopeResolver::GetRandomBits() const
+{
+	if (this->st == nullptr) return 0;
+
+	uint32 bits = this->st->random_bits;
+	if (this->tile != INVALID_TILE && Station::IsExpected(this->st)) {
+		bits |= Station::From(this->st)->GetRoadStopRandomBits(this->tile) << 16;
+	}
+	return bits;
+}
+
+uint32 RoadStopScopeResolver::GetTriggers() const
+{
+	return this->st == nullptr ? 0 : this->st->waiting_triggers;
+}
+
+uint32 RoadStopScopeResolver::GetVariable(byte variable, uint32 parameter, bool *available) const
+{
+	auto get_road_type_variable = [&](RoadTramType rtt) -> uint32 {
+		RoadType rt;
+		if (this->tile == INVALID_TILE) {
+			rt = (GetRoadTramType(this->roadtype) == rtt) ? this->roadtype : INVALID_ROADTYPE;
+		} else {
+			rt = GetRoadType(this->tile, rtt);
+		}
+		if (rt == INVALID_ROADTYPE) {
+			return 0xFFFFFFFF;
+		} else {
+			return GetReverseRoadTypeTranslation(rt, this->roadstopspec->grf_prop.grffile);
+		}
+	};
+
+	switch (variable) {
+		/* View/rotation */
+		case 0x40: return this->view;
+
+		/* Stop type: 0: bus, 1: truck, 2: waypoint */
+		case 0x41:
+			if (this->type == STATION_BUS) return 0;
+			if (this->type == STATION_TRUCK) return 1;
+			return 2;
+
+		/* Terrain type */
+		case 0x42: return this->tile == INVALID_TILE ? 0 : GetTerrainType(this->tile, TCX_NORMAL); // terrain_type
+
+		/* Road type */
+		case 0x43: return get_road_type_variable(RTT_ROAD);
+
+		/* Tram type */
+		case 0x44: return get_road_type_variable(RTT_TRAM);
```

Comment:
I'd be tempted to use fallthrough & a `RoadTramType rtt = 0x43 ? RTT_ROAD : RTT_TRAM` variable personally, but eh